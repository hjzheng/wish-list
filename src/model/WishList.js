import { types } from 'mobx-state-tree';

// const data = {
//     "name": "Chronicles of Narnia Box Set",
//     "price": 28.73,
//     "image": "https://images-na.ssl/....png"
// }

export const WishListItem = types.model('WishListItem', {
    name: types.string,
    price: types.number,
    image: types.optional(types.string, '')
})
.actions(self => ({
    changeName(newName) {
        self.name = newName
    },
    changePrice(newPrice) {
        self.price = newPrice;
    },
    changeImage(newImage) {
        self.image = newImage;
    },
}))

export const WishList = types.model('WishList', {
    items: types.optional(types.array(WishListItem), [])
})
.actions(self => ({
    addItem(item) {
        self.items.push(item);
    }
}))
.views(self => ({
    get totolPrice() {
        return self.items.reduce((sum, item) => sum + item.price, 0);
    }
}))