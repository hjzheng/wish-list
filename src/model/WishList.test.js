import { WishListItem, WishList } from './WishList';
import { getSnapshot, onSnapshot, onPatch} from "mobx-state-tree"
import { reaction } from 'mobx';

it('can create WishLisItem', () => {
    const item = WishListItem.create({
        "name": "Chronicles of Narnia Box Set",
        "price": 28.73,
        "image": "https://images-na.ssl/....png"
    })

    expect(item.name).toBe("Chronicles of Narnia Box Set");
    expect(item.price).toBe(28.73);

    item.changeName('Harry Zheng');

    expect(item.name).toBe('Harry Zheng');
})


it('can create WishList', () => {
    const list = WishList.create({
        items: [{
            "name": "Chronicles of Narnia Box Set",
            "price": 28.73,
        }]
    })

    expect(list.items[0].name).toBe("Chronicles of Narnia Box Set");
    expect(list.items.length).toBe(1);
})

it('add new item', () => {
    const list = WishList.create()
    list.addItem(WishListItem.create({
        name: 'Harry Zheng',
        price: 22
    }))

})

it('use getSnapshot', () => {
    const list = WishList.create()
    
    list.addItem(WishListItem.create({
        name: 'Xbox One',
        price: 3600
    }))

    expect(getSnapshot(list)).toEqual({
        items: [{
            name: 'Xbox One',
            price: 3600,
            image: ''
        }]
    })
    
    expect(getSnapshot(list)).toMatchSnapshot()

})

it('use onSnapshot', () => {
    const list = WishList.create()

    const states = []

    onSnapshot(list, snapshot => {
        states.push(snapshot)
    })
    
    list.addItem(WishListItem.create({
        name: 'PS 4',
        price: 2800
    }))

    expect(states).toMatchSnapshot()
})

it("use onPatch", () => {
    const list = WishList.create()
    const patches = []
    onPatch(list, patch => {
        patches.push(patch)
    })

    list.addItem({
        name: "Chesterton",
        price: 10
    })

    list.items[0].changeName("Book of G.K. Chesterton")

    expect(patches).toMatchSnapshot()
})

it('computed total price', () => {
    const list = WishList.create({
        items: [
            {
                name: "PS4",
                price: 2800
            },
            {
                name: "Xbox One with kinect",
                price: 3600
            },
        ]
    })

    expect(list.totolPrice).toBe(6400);

    let change = 0;

    reaction(() => list.totolPrice, () => change++)

    expect(change).toBe(0);

    list.items[0].changePrice(2900);

    expect(change).toBe(1);

    list.items[0].changeName('PS4 Pro');

    expect(change).toBe(1);
})