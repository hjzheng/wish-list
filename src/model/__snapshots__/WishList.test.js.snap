// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`use getSnapshot 1`] = `
Object {
  "items": Array [
    Object {
      "image": "",
      "name": "Xbox One",
      "price": 3600,
    },
  ],
}
`;

exports[`use getSnapshot 2`] = `
Object {
  "items": Array [
    Object {
      "image": "",
      "name": "Xbox One",
      "price": 3600,
    },
  ],
}
`;

exports[`use onPatch 1`] = `
Array [
  Object {
    "op": "add",
    "path": "/items/0",
    "value": Object {
      "image": "",
      "name": "Chesterton",
      "price": 10,
    },
  },
  Object {
    "op": "replace",
    "path": "/items/0/name",
    "value": "Book of G.K. Chesterton",
  },
]
`;

exports[`use onSnapshot 1`] = `
Array [
  Object {
    "items": Array [
      Object {
        "image": "",
        "name": "PS 4",
        "price": 2800,
      },
    ],
  },
]
`;
